### PR TITLE
Change caret style

### LINKED ISSUE #3936: Can't see cursor in caret mode on some pages
Version info: http://paste.the-compiler.org/view/0a29b429

On most pages I can't see the cursor in caret mode. When I can see it, it's very hard to actually see it as it seems to be a 1 pixel wide line that's the same colour as most of the text.

Is there some way of changing the style? Maybe a blinking block would be easier to see.

### PR DESCRIPTION
**Pull request by slackhead**
This changes some of the style of the caret:

Use inherit rather than set colours so whatever is in the parent element is used.
min-width of 0.6em uses roughtly the whole character width, and it resizes in proportion with zoom and/or min-font-size. To use a vertical line it could be changed to 0.2em perhaps (untested).
mix-blend-mode inverts the text colour.
blink helps to make it stand out more
background could be background-color. Not sure yet what difference it makes.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3940)

*edit by @The-Compiler to link issues: Fixes #3936*

### CODE REVIEW

**Comment by slackhead** [comment]
Yeah I noticed the same thing with the settings page.

**Comment by slackhead** [comment]
I've removed the 'inherit' styles.

Colours will be inherited automatically and the invert filter will do its job on those.

I tested on the settings page, a page with a black background and a white background.

I also tested setting min-width to 0.2 to produce a vertical bar. It's thicker than the original vertical bar so a lot easier to see.

File: qutebrowser/javascript/caret.js [hunk 1]
```diff
@@ -860,11 +860,20 @@ window._qutebrowser.caret = (function() {
...
+            "}";
+        const blink = "@keyframes blink {" +
+            "50% { visibility: hidden; }" +
             "}";
         const node = document.createElement("style");
         node.innerHTML = style;
         document.body.appendChild(node);
+        const node2 = document.createElement("style");
+        node2.innerHTML = blink;
+        document.body.appendChild(node2);
```
Comment (The-Compiler) [inline qutebrowser/javascript/caret.js]: Why not have this in the same node than the rest of the CSS?

**Comment by The-Compiler** [comment]
Hmm, I just tried this and I don't see a cursor anymore at all.

**Comment by slackhead** [comment]
That would represent two nodes but you can only append one at a time.

I need to fix the colour because it's stopped working without inherit now.

**Comment by The-Compiler** [comment]
Yeah, but why do we need two style-nodes at all? Why not have the entire style in a single node?

**Comment by slackhead** [comment]
As far as I know there's no way of including the @keyframes inside the main chunk, so it has to be a separate node.

**Comment by jgkamat** [comment]
The new commit seems to work fine on unstyled pages, but I'm not sure how I feel about the blinking. Since blinking continues even when moving around, it's hard to see where the cursor is. I'm not sure if we should just turn it off or have an option for it (if we can't make it stay solid while moving).

**Comment by slackhead** [comment]
I've tried adding keydown and keyup functions to change the blinking, but they don't work when using movement keys, so I guess they are intercepted elsewhere:

        window.onkeydown = function() {
            const element = CaretBrowsing.caretElement;
            element.ClassName = "CaretBrowsing_Caret";
            element.style.animationDuration = 1;
        };

Adding an alert() in there shows it's working, but it only pops up with a key that isn't bound, not for j,k,h,l etc.

**Comment by The-Compiler** [comment]
You should be able to modify `CaretBrowsing.move` and `CaretBrowsing.moveToBlock`

also cc'ing @artur-shaik here since he knows the caret code better than I do :wink:

**Comment by slackhead** [comment]
Added CaretBrowsing.stopAnimation() and CaretBrowsing.startAnimationt()

stopAnimation() is called from CaretBrowsing.move and CaretBrowsing.moveToBlock and sets a 1 sec timer to animationStart() which avoids some flickering from continually stopping/starting.

There are still some issues with text colour on some pages, which I'm still looking at. For some reason text colour seems to be ignored and is set to the same colour as the background so that you end up with a flashing grey block. It does this on paragraph text in the local help pages and some sites like w3schools (irony?).

**Comment by slackhead** [comment]
OK, I am admitting defeat on finding why the cursor fg colour is sometimes grey. 

I've changed the min-width to 0.2em because of that, because grey on grey doesn't really work with a block cursor.

It is still more visible than before, and the width can be changed via a user stylesheet if needed.

**Comment by slackhead** [comment]
To use a block cursor a user stylesheet can contain:

.CaretBrowsing_Caret {
	min-width: 0.6em;
	filter: invert(50%);
}

File: qutebrowser/javascript/caret.js [hunk 2]
```diff
@@ -859,12 +859,25 @@ window._qutebrowser.caret = (function() {
...
+        const blink = "@keyframes blink {" +
+            "50% { visibility: hidden; }" +
             "}";
         const node = document.createElement("style");
         node.innerHTML = style;
         document.body.appendChild(node);
+
+        const node2 = document.createElement("style");
+        node2.innerHTML = blink;
+        document.body.appendChild(node2);
```
Comment (The-Compiler) [inline qutebrowser/javascript/caret.js]: I still don't understand why a second element is needed here. I tried this and it seems to work just fine:

```diff
diff --git a/qutebrowser/javascript/caret.js b/qutebrowser/javascript/caret.js
index 800a27735..3a4272984 100644
--- a/qutebrowser/javascript/caret.js
+++ b/qutebrowser/javascript/caret.js
@@ -867,17 +867,13 @@ window._qutebrowser.caret = (function() {
             "  mix-blend-mode: difference;" +
             "  filter: invert(85%);" +
             "  animation: blink 1s step-end infinite;" +
-            "}";
-        const blink = "@keyframes blink {" +
-            "50% { visibility: hidden; }" +
+            "}" +
+            "@keyframes blink {" +
+            "  50% { visibility: hidden; }" +
             "}";
         const node = document.createElement("style");
         node.innerHTML = style;
         document.body.appendChild(node);
-
-        const node2 = document.createElement("style");
-        node2.innerHTML = blink;
-        document.body.appendChild(node2);
     };
 
     /**
```

File: qutebrowser/javascript/caret.js [hunk 3]
```diff
@@ -932,7 +945,7 @@ window._qutebrowser.caret = (function() {
      */
     CaretBrowsing.setCaretElementNormalStyle = function() {
         const element = CaretBrowsing.caretElement;
-        element.className = "CaretBrowsing_Caret";
+        element.ClassName = "CaretBrowsing_Caret";
```
Comment (The-Compiler) [inline qutebrowser/javascript/caret.js]: Why this change? `className` seems to be the canonical spelling: https://developer.mozilla.org/en-US/docs/Web/API/Element/className

File: qutebrowser/javascript/caret.js [hunk 4]
```diff
@@ -1242,6 +1259,17 @@ window._qutebrowser.caret = (function() {
         CaretBrowsing.updateIsCaretVisible();
     };
 
+    CaretBrowsing.startAnimation = function() {
+        CaretBrowsing.caretElement.style.animationIterationCount = 'infinite';
```
Comment (The-Compiler) [inline qutebrowser/javascript/caret.js]: please use double quotes

**Comment by slackhead** [comment]
It seems that qt 5.7.1 doesn't support mix-blend-mode and possibly not filter and var() too.

I've simplified it by removing those and setting background-color back to #000 as it was.

Anyone wanting to use the removed features can add them back in in c.content.user_stylesheets. eg:

.CaretBrowsing_Caret {
	min-width: 0.6em;
	--inherited-color: inherit;
	background-color: var(--inherited-color, #000);
	color: var(--inherited-color, #000);
	mix-blend-mode: difference;
	filter: invert(45%);
}

**Comment by The-Compiler** [comment]
Looking at [caniuse](https://caniuse.com/) it apparently does support [mix-blend-mode](https://caniuse.com/#feat=css-mixblendmode) and [var()](https://caniuse.com/#feat=css-variables), but [filter](https://caniuse.com/#feat=css-filters) was prefixed (`-webkit-filter`) back then.

I went back to the old style and added the prefix conditionally, and now it seems to work:

https://github.com/qutebrowser/qutebrowser/blob/1541088e76478e532bb6f6e112b81707d47548e7/qutebrowser/javascript/caret.js#L870-L892

(I also switched to ES6 template strings for readability)

Thanks! Definitely a lot more visible this way.

**Comment by slackhead** [comment]
I'm finding that the caret it light grey and very hard to see on ebay.co.uk pages in certain lines of text. Setting background-color back to black, color to white and filter to 'none' works better on those pages.

May I suggest keeping it as it was before the prefix conditional was added, and leaving it for users to modify it if they want to? Black and white is very easy to see.

Maybe I could write up a few notes on which settings are useful for that.

**Comment by The-Compiler** [comment]
No problem! Always happy to find a proper solution for issues :smile:

But with the colors hardcoded like that, wouldn't the cursor be invisible on pages with a black background?

**Comment by slackhead** [comment]
This is true. At the moment though it isn't working for white text on a black background very well. See http://r0t.uk 

In 5.7.1 that
mix-blend-mode: difference;
line isn't working for me. I don't see the caret at all unless I delete it.
The blending doesn't seem to be working. If you change the % to 45% or so the caret should be a greyish colour, but it's black for me.

I'll keep working on it until I get some better results.

**Comment by recallthis** [comment]
Is it possible using these methods to highlight the whole line that the caret is on? I mean either, the full width of the horizontal space implied by the height of the caret, or just the full line of text (with a different color or shade to text selection) 

Like in vim, emacs, ranger etc, a visual aid and anchor for line selection.
